### DISCUSSION TITLE
(Pedantic) thoughts on naming conventions

### DISCUSSION THREAD

**Discussion by dflorey** [firebase/firebase-js-sdk#4563] [General]
While migrating my app to the new modulized sdk I've been running into some name clashes, espcially for the doc() method that clashes with my "doc" variable in the callback. While this is of course just a minor annoyance I've been thinking about the naming conventions in the sdk a bit.
I'd prefer a consistent naming scheme like this (which would also reduce the number of clashes with variable names that I guess are widely used with the old sdk):
Using either full or abbreviated method names, so either
collection(...) document(...) onSnapshot(...) setDocument(...) etc.
or 
col(...) doc(...) onSnap(...) setDoc(...)

I'd prefer the full names, as it is nicer to read and I guess it will also not affect the generated bundle size.
The "Common use case" would then be much nicer to read as 

const docSnap = await getDoc(docRef); // <- Using docSnap as doc would clash with the imported method

would become

const doc = await getDocument(docRef)

etc.

A new SDK version is a great opportunity to start with fresh and consistent naming schemes, so even though it may be just a matter of taste I think it is worth to provide some feedback.

**Comment by merlinnot** [comment]
Agreed, it was always surprising for me that we have `.collection` and `.doc`. I'd prefer full names as well, especially that it looks more tidy when used alongside type definitions, which are all full names.

**Comment by mbleigh** [comment]
I vaguely remember our original conversations about this. We ended up here because "doc" is a universally recognized shorthand for "document" (no risk of confusion) and was also something that was going to end up getting typed a lot. "Collection" does not have a clear analogue -- "col" is a common abbreviation for "column" and "coll" doesn't really pop. "Snap" is *reasonably* clear for Snapshot, but it's not quite as universal.

Would folks feel better (as a counterproposal without any thought behind it) if `Document` were universally renamed to `Doc` for consistency of a single term?

**Reply by dflorey** [reply]
If I'd be in charge of the universe I'd have chosen a different name in the first place instead of doc (especially since Google has a product called Docs). Something like entity, item, record or whatever makes sense in a collection without implying any special use (doc / file is somehow already taken).
As we are stuck with the document term I'd vote for the long version (document(), setDocument() etc.) as we have auto-completion anyway (so typing a few more chars is not an issue) and it would free up "doc" as an available variable name which makes sense in the callback.
I hope my feedback does not sound harsh (not a native speaker), I really enjoy the new sdk and I'm just trying to provide some feedback to make it even better :-)

**Reply by mbleigh** [reply]
It's appreciated feedback, you don't sound harsh. We're going to talk more about this internally. FWIW (though this is largely due to my long-time use of the Firebase Realtime Database) I tend to use "ref" in my variable names since docs and collections are both references:

```ts
const userRef = doc(db, "users/mbleigh");
```

**Reply by dflorey** [reply]
Ok, so you get a few more comments on the naming :-)
https://modularfirebase.web.app/common-use-cases/firestore/#crud-operations
In the first example "docSnap" is used as variable name, even though the document has not been fetched via onSnapshot(), so this has been a bit misleading for me at first sight.
If you scan through the other examples you'll find similar confusing variable names. I think for the docs (I mean the documentation) it would be nice to carefully pick the names as it will be easier for new users to not get confused.
Using "snapshot" only for results from onSnapshot() and "document" or "documents" for results from getDoc() or getDocs() would make it more obvious.

**Reply by mbleigh** [reply]
Actually sounds like we do have some clarifying to do here! The value returned from an awaited `getDoc()` is a `DocumentSnapshot`, just like the value of an `onSnapshot` for a doc is a `DocumentSnapshot`. In both cases, you are getting a snapshot of the document at a point in time. There is no such thing as a `Document` in the SDK -- there is a `DocumentReference` which is a pointer to a specific document, and a `DocumentSnapshot` which contains the contents of the document at a particular point in time.

So `docSnap` is correct for `getDoc()` and `onSnapshot()` alike, as in both cases you get the same type back.

**Reply by dflorey** [reply]
Hm, so what does the term "Snapshot" mean then? Sounds like "onSnapshot()" is something like "onChange()" then?

**Reply by mbleigh** [reply]
Huh, I never thought of it before but you're right that `onSnapshot` makes it sound like a verb (a snapshot is happening). I think it would more accurately be `onNewSnapshot`, but yeah it's effectively `onChange` (although I think the reason that we don't call it that is because our event listeners *also* trigger immediately with the current value even if it hasn't changed).

Lots of subtlety here and it's definitely something that I think many of us are taking for granted, so thanks for the interesting discussion!

**Reply by dflorey** [reply]
Probably it is because I'm (relatively) new to firebase and I've used the "old" Datastore before. I've then been reading about the new datastore (aka firestore) and the docs are talking about "Documents" only:
https://firebase.google.com/docs/firestore/data-model
So it is not clear at first sight what a "snapshot" is - as this term is never introduced. It gets kind of clear how to use it though, but to me it would be more easy to understand if getDoc() aka getDocument() would just return a Document and onSnapshot() would be onDocument() or something along those lines...
Anyway, I think a few variable names in the new docs could be aligned to improve the readability (without changing any of the given mwthod/type names in the sdk itself)

**Reply by akauppi** [reply]
@mbleigh Hey - you mentioned Realtime Database!

Would it be too much to ask to consider calling that product `realtime` in the API?  Having `database` is simply confusing, because also Firestore is one.

**Reply by mbleigh** [reply]
That would be a huge change across many platforms and I don't think it likely to be something we're looking to tackle. In general "Firebase Realtime Database" is often shortened to "Firebase Database" and "Firestore" is always just called "Firestore" (or sometimes "Cloud Firestore" but we're trying to use that less going forward I believe).

Also, they're both databases but Firestore is *also* realtime 😄

**Reply by akauppi** [reply]
@mbleigh Thanks for the comment. Seeing you point out Firestore as *realtime* raises my eyebrows ;) a bit. I haven’t come to the realtime needs of my web app, yet, but thought I would need to do it with RTDB (in addition to Firestore in the same project, which cannot be done), or some Websockets like mechanism.

My use case there is to share temporary information across a group of users; think collaborative editing. The data does not need to be persisted, just distributed.

*This is off topic, so muting myself.*
